torchvision/prototype/transforms/functional/_geometry.py

```diff
@@ -1,21 +1,51 @@
-from typing import Tuple, List, Optional
+from typing import Tuple, List, Optional, TypeVar, Any
 
 import torch
-from torchvision.transforms import (  # noqa: F401
-    functional as _F,
-    InterpolationMode,
-)
+from torchvision.prototype import features
+from torchvision.transforms import functional as _F, InterpolationMode
+
+from ._meta_conversion import convert_bounding_box_format
+from .utils import dispatch
+
+T = TypeVar("T", bound=features.Feature)
+
+
+@dispatch
+def horizontal_flip(input: T) -> T:
+    """ADDME"""
+    pass
+
 
 horizontal_flip_image = _F.hflip
+horizontal_flip.register(features.Image, horizontal_flip_image)
 
 
-def horizontal_flip_bounding_box(bounding_box: torch.Tensor, *, image_size: Tuple[int, int]) -> torch.Tensor:
-    bounding_box = bounding_box.clone()
+def horizontal_flip_bounding_box(
+    bounding_box: torch.Tensor, *, format: features.BoundingBoxFormat, image_size: Tuple[int, int]
+) -> torch.Tensor:
+    bounding_box = convert_bounding_box_format(
+        bounding_box, old_format=format, new_format=features.BoundingBoxFormat.XYXY, copy=True
+    )
     bounding_box[..., (0, 2)] = image_size[1] - bounding_box[..., (2, 0)]
-    return bounding_box
+    return convert_bounding_box_format(bounding_box, old_format=features.BoundingBoxFormat.XYXY, new_format=format)
 
 
-_resize_image = _F.resize
+@horizontal_flip.implements(features.BoundingBox)
+def _horizontal_flip_bounding_box(input: features.BoundingBox) -> torch.Tensor:
+    return horizontal_flip_bounding_box(input, format=input.format, image_size=input.image_size)
+
+
+@dispatch
+def resize(
+    input: T,
+    *,
+    size: List[int],
+    interpolation: InterpolationMode = dispatch.FEATURE_SPECIFIC_DEFAULT,  # type: ignore[assignment]
+    max_size: Optional[int] = None,
+    antialias: Optional[bool] = None,
+) -> T:
+    """ADDME"""
+    pass
```

Comment:
> Do we really need to delay the registration into multiple commands vs RAII (resize is created with provided resize_image, resize_bbox etc at construction time) ?

*We* don't need to, no. But we would have to keep this mechanism if someone outside of `torchvision` wants to register a custom `Feature` kernel for a transform. Imagine I have `MyAwesomeFeature` and a kernel `resize_my_awesome_feature` how would I add it to the default `F.resize` dispatch if we don't have an option to register at runtime?

> What would happen if signature of `resize` is different from `resize_image`, `resize_segmentation_mask` etc ? and to what corresponds the signature of `resize` ?

Just like the body of the dispatcher function, the signature is only used for the documentation. If it diverges from the kernel signatures, we should have explanations in the docstring what is happening.

Reply:
> We don't need to, no. But we would have to keep this mechanism if someone outside of torchvision wants to register a custom Feature kernel for a transform. Imagine I have MyAwesomeFeature and a kernel resize_my_awesome_feature how would I add it to the default F.resize dispatch if we don't have an option to register at runtime?

If we think to expose this as public API, we have to test it better, write docs etc. 
Yes, feature like extending resize to make it work on other types would be nice but then we have to ensure which scope we provide. Can user override `resize_image` for example ? Can we register on other non-tensor types, e.g. `str`:
```python
@resize.implements(str)
def resize_str(image, *args, **kwargs):
    print("Call resize_str")
    return image[1:-1]
```
(Today, this does not error but raises an exception only on call)
How about parity then ? If we provide publicly add operation (`dispatch.register`), then we may want to provide removal operation, e.g `dispatch.deregister`. I do not see any application of that at the moment...

Being careful today, I'd make `register` as private method and do not expose right now for the users.

Reply:
@vfdev-5 

> If we think to expose this as public API, we have to test it better, write docs etc.

Of course, but that has time until we actually expose this outside of `torchvision.prototype`.

> Yes, feature like extending resize to make it work on other types would be nice but then we have to ensure which scope we provide.

I don't think this is nice-to-have, but rather integral. Our transforms should be flexible enough to cater to the needs of 90% of our user base without them using hacky workarounds. To give an example, Franciscos new project already needs this if they want to adopt our new transforms.

> Can we register on other non-tensor types, e.g. `str`

Nope, not for now. The output wrapping into a `Feature` already implied this, but I will make it explicit with an error. This way, we have instant feedback.

> How about parity then ? If we provide publicly add operation (`dispatch.register`), then we may want to provide removal operation, e.g `dispatch.deregister`. I do not see any application of that at the moment...

I think that is fair game to wait for someone actually requesting this. In the current setup if you call `register` again with an already registered feature type, the old entry is simply overwritten. Thus, if the user wants to have a custom kernel this is already supported. 

I don't see a use case for de-registering feature types. If we had such a functionality, calling a dispatcher with a de-registered feature type would result in an error. Thus, it should be much simpler to not call the dispatcher in the first place with whatever feature type you want to de-register.

---

@datumbox 

> This is a potential expansion feature but it shouldn't be our target for the first version.

See my answer to @vfdev-5 above.

> I agree that we should provide these at construction time as arguments to the dispatch annotation to simplify code and make it less verbose.

I don't think it would simplify the implementation or make it less verbose. Compare

```py
@dispatch
def resize(...):
    ...


resize.register(resize_image, features.Image, pil_kernel=_F.resize)
resize.register(resize_segmentation_mask, features.SegmentationMask)
```

with

```py
@dispatch(
    {
        features.Image: resize_image,
        features.SegmentationMask: resize_segmentation_mask,
    },
    pil_kernel=_F.resize,
)
def resize(...):
    ...
```

Plus, we would need to have extra indication if the callable we pass handles the output wrapping or not. With the `register` approach we can simply set `wrap_output=False` which automatically only applies to the current callable.

Reply:
Marking this thread as resolved since the latest commits (hopefully) bring the `@dispatch` decorator more in the direction of your suggestions.